edpop_explorer/readers/stcn.py

```diff
@@ -1,104 +1,241 @@
-from rdflib import Graph, Namespace, URIRef
-from rdflib.term import Node
+from rdflib import URIRef
 from typing import List, Optional, Tuple
 
-from edpop_explorer import Field, BIBLIOGRAPHICAL
-from edpop_explorer.fields import LanguageField
-from edpop_explorer.sparqlreader import (
-    SparqlReader, BibliographicalRDFRecord
-)
-
-
-def _get_properties_from_iri(iri: str, properties: List[Node]) -> \
-        Tuple[List[Node], Graph]:
-    '''Get the first objects of the requested properties of a certain IRI
-    as strings.'''
-    subject_graph = Graph()
-    subject_graph.parse(iri)
-    objects: List[Node] = []
-    for prop in properties:
-        for obj in subject_graph.objects(URIRef(iri), prop):
-            objects.append(obj)
-    return objects, subject_graph
-
-
-class STCNReader(SparqlReader):
-    endpoint = 'http://data.bibliotheken.nl/sparql'
-    filter = '?s schema:mainEntityOfPage/schema:isPartOf ' \
-        '<http://data.bibliotheken.nl/id/dataset/stcn> .'
-    name_predicate = '<http://schema.org/name>'
+from edpop_explorer import Field, BIBLIOGRAPHICAL, BibliographicalRecord, LocationField, BIOGRAPHICAL, \
+    BiographicalRecord
+from edpop_explorer.cerl import CERLReader
+from edpop_explorer.fields import LanguageField, ContributorField
+
+
+def _remove_markup(input_str: str) -> str:
+    """Remove STCN-specific markup"""
+    return input_str.replace('`IT`', '').replace('`LO`', '')
+
+
+def safeget(dictionary: Optional[dict], attribute_chain: tuple, first: bool = False):
+    attribute = attribute_chain[0]
+    if dictionary is None or attribute not in dictionary:
+        return None
+    value = dictionary[attribute]
+    if first and isinstance(value, list):
+        value = value[0]
+    if len(attribute_chain) == 1:
+        return value
+    else:
+        return safeget(value, attribute_chain[1:], first)
+
+
+class STCNBaseReader(CERLReader):
+    """STCN uses the same search API for its bibliographical records and
+    its biographical records (persons and publishers/printers), but the
+    data format as well as detail pages are different. This base class
+    builds on CERLReader and adds the API URL."""
+    API_URL = 'https://data.cerl.org/stcn/_search'
+
+
+class STCNPersonsReader(STCNBaseReader):
+    API_BY_ID_BASE_URL = 'https://data.cerl.org/stcn_persons/'
+    LINK_BASE_URL = 'https://data.cerl.org/stcn_persons/'
+    CATALOG_URIREF = URIRef(
+        'https://edpop.hum.uu.nl/readers/stcn'
+    )
+    IRI_PREFIX = "https://edpop.hum.uu.nl/readers/stcn-persons/"
+    READERTYPE = BIOGRAPHICAL
+    SHORT_NAME = "STCN Persons"
+    DESCRIPTION = "National bibliography of The Netherlands until 1801 – persons"
+
+    @classmethod
+    def transform_query(cls, query) -> str:
```

Comment:
The reason I did this is that `_perform_query` is also used by the `get_by_id` method in the superclass, which does not require an instance of the class because it is a one-time fetch. My idea was that an instance is only required when querying/fetching a range, because then the results are saved in the instance and the fetch method can be run multiple times.

Would you rather change that design? I could either make `get_by_id` an instance method (but then I have to do that everywhere in the EDPOP Explorer) or change the class method that don't use `cls` into static methods.